### PR TITLE
Hard coded startup

### LINKED ISSUE #422: Replace filesystem startup sequence with hardcoded rate-group-based startup
Short summary
Replace the file-based startup.seq with a hardcoded startup flow in StartupManager: wait 45 minutes after boot, then enable radio transmit. 

Motivation and user story
As a flight software developer, I want startup logic baked into firmware and driven by rate groups, so that boot behavior is predictable, does not depend on a sequence file, and is not affected by RTC issues or TIME_SET.

Today StartupManager dispatches /seq/startup.bin via cmdSeq on first boot. startup.seq mixes RTC-relative timestamps with many commands (antenna deploy, detumble, LoRa config, safe mode exit, etc.). A simpler mission need is: wait 45 minutes, then turn on transmit (similar to radio_enter_safe.seq).

Proposed implementation
Refactor StartupManager to run a minimal hardcoded state machine on the existing 1 Hz rate group (startupManager.run)

Keep startup sequence running since we can edit that

boot, start a 45-minute countdown using uptime / rate group ticks 
After the wait, enable radio transmit via a new output port wired to the radio component (instead of dispatching a sequence file).

### PR DESCRIPTION
**Pull request by ineskhou**
# Add Hard Coded Startups

## Description

This gives a hardcoded startup for the startup component. It will count 45 minutes worth of tiks, and enable the radio from there. This will allow us to bypass a broken filesystem.

There are some port connections that have been added here.

Sequence tracking (any sequence running)

      cmdSeq.seqStartOut -> startupManager.startupsequenceStarted
      cmdSeq.seqDone -> startupManager.startupCompleteSequence

      payloadSeq.seqStartOut -> startupManager.payloadSequenceStarted
      payloadSeq.seqDone -> startupManager.payloadCompleteSequence

      # seqDone is owned by ModeManager; completion is forwarded via sequenceDoneNotify
      safeModeSeq.seqStartOut -> startupManager.safeModeSequenceStarted

And this is made for the passthrough for the last sequencer

      modeManager.runSequence -> safeModeSeq.seqRunIn
      safeModeSeq.seqDone -> modeManager.completeSequence
      modeManager.sequenceDoneNotify -> startupManager.safeModeCompleteSequence

Tracking Lora Turning on

   # --- Radio ever enabled this boot? ---
      lora.loraFirstStart -> startupManager.loraFirstStart

Give startup manager the abillity to turn radio on

      # StartupManager drives LoRa TX enable/disable around quiescence
      startupManager.enableTransmit -> lora.enableTransmit
      startupManager.disableTransmit -> lora.disableTransmit

Hardcoded TX enable: 1 Hz countdown (TRANSMIT_ENABLE_TICKS); on expiry asserts enableTransmit
HardCodedStartup.h: compile flag DEFAULT_STARTUP_VALUE (0 = off for ground, 1 = on for flight)
HardcodedRadioEnable event when that path fires
Defer if a sequence is running (m_active_sequences refcount + pending flag); enable after last seq completes
Skip if radio already on (loraFirstStart → m_lora_ever_on)
Split old start/done ports into per-sequencer ports + loraFirstStart / enable/disable outs

## Related Issues/Tickets

https://github.com/Open-Source-Space-Foundation/proves-core-reference/issues/422 

## How Has This Been Tested?

- [ ] Unit tests
- [ ] Integration tests
- [ ] Z Tests
- [X] Manual testing (describe steps)

Ran the sequence for 45 mins and obsereved the radio turning on

## Checklist

- [ ] Written detailed sdd with requirements, channels, ports, commands, telemetry defined and correctly formatted and spelled
- [ ] Have written relevant integration tests and have documented them in the sdd
- [ ] Have done a code review with
- [ ] Have tested this PR on every supported board with correct board definitions

### CODE REVIEW

**Comment by ineskhou** [comment]
Set the Tick Timer to 100 seconds:

Screenshot 2026-07-17 at 11 44 03 PM

Screenshot 2026-07-17 at 11 44 11 PM

Set the Tick Timer for 45 minutes + 100 seconds of lag
Screenshot 2026-07-18 at 12 36 44 AM


Screenshot 2026-07-18 at 12 36 27 AM

File: PROVESFlightControllerReference/Components/StartupManager/StartupManager.fpp [hunk 1]
```diff
@@ -60,6 +70,9 @@ module Components {
         @ File to store the boot count
         param BOOT_COUNT_FILE: string default "/boot_count.bin"
 
+        @ 1 Hz run ticks before enabling LoRa transmit (45*60 + 100 margin)
+        param TRANSMIT_ENABLE_TICKS: U32 default 100
+
```
[the remark this answers is not in the document]
Comment (ineskhou) [reply PROVESFlightControllerReference/Components/StartupManager/StartupManager.fpp:90]: I think I must have pushed my testing code, let me correct this!

File: PROVESFlightControllerReference/Components/StartupManager/StartupManager.hpp [hunk 2]
```diff
@@ -98,12 +98,13 @@ class StartupManager final : public StartupManagerComponentBase {
...
-    U32 m_stored_sequence;         //!< Stored sequence number for delayed response
-    std::atomic<bool> m_waiting;   //!< Indicates if waiting for quiescence
-    Fw::String m_sequence_file;    //!< The filepath for the sequence last initiated
+    Fw::Time m_quiescence_start;      //!< Time of the start of the quiescence wait
+    FwOpcodeType m_stored_opcode;     //!< Stored opcode for delayed response
+    FwSizeType m_boot_count;          //!< Current boot count
+    U32 m_stored_sequence;            //!< Stored sequence number for delayed response
+    std::atomic<bool> m_waiting;      //!< Indicates if waiting for quiescence
+    Fw::String m_sequence_file;       //!< The filepath for the sequence last initiated
+    U32 m_transmit_enable_ticks = 0;  //!< Remaining 1 Hz ticks until hard-coded transmit enable
```
[the remark this answers is not in the document]
Comment (ineskhou) [reply PROVESFlightControllerReference/Components/StartupManager/StartupManager.hpp]: Based on the way we've been naming variables in F Prime will keep these to be consistant

File: docs-site/components/RtcManager.md [hunk 3]
```diff
@@ -97,31 +143,51 @@ classDiagram
...
+            + configure(dev: const device*) void
+
+            - timeGetPort_handler(portNum: FwIndexType, time: Fw::Time&) void
+
+            - TIME_SET_cmdHandler(opCode: FwOpcodeType, cmdSeq: U32, t: Drv::TimeData) void
+            - ALARM_SET_cmdHandler(opCode: FwOpcodeType, cmdSeq: U32, t: Drv::TimeData) void
+            - ALARM_CANCEL_cmdHandler(opCode: FwOpcodeType, cmdSeq: U32, ID: U16) void
+            - ALARM_LIST_cmdHandler(opCode: FwOpcodeType, cmdSeq: U32) void
+
+            - static_alarm_callback_t(dev: const device*, id: uint16_t, user_data: void*) void$
```
[the remark this answers is not in the document]
Comment (Mikefly123) [reply docs-site/components/RtcManager.md:165]: Unrelated to this PR.

File: docs-site/components/RtcManager.md [hunk 4]
```diff
@@ -15,6 +18,32 @@ The RTC Manager component interfaces with the Real Time Clock (RTC) to provide t
...
+    - Cancels the present alarm if it is present on the system
+    - Emits a `AlarmCanceled` event with the previous time if the alarm is canceled successfully
+    - Emits a `AlarmNotCanceled` event if the alarm is not canceled successfully
+
+#### `ALARM_LIST` Command Usage
+1. A ground station sends a `ALARM_LIST` command
+2. On each command, the component:
+    - Validates that an alarm exists
+    - Emits 'AlarmNotSet' if no alarm is present
+    - Emits 'AlarmSet' if an alarm is present and returns it's information to the ground station
```
[the remark this answers is not in the document]
Comment (Mikefly123) [reply docs-site/components/RtcManager.md:45]: Unrelated to this PR.

File: docs-site/components/RtcManager.md [hunk 5]
```diff
@@ -15,6 +18,32 @@ The RTC Manager component interfaces with the Real Time Clock (RTC) to provide t
...
+    - Emits a `AlarmNotSet` event if the alarm is not set successfully
+
+#### `ALARM_CANCEL` Command Usage
+1. A ground station sends a `ALARM_CANCEL` command with the desired alarm ID to cancel
+2. On each command, the component:
+    - Validates that an alarm exists
+    - Emits 'AlarmNotCanceled' if an alarm is already present
+    - Cancels the present alarm if it is present on the system
+    - Emits a `AlarmCanceled` event with the previous time if the alarm is canceled successfully
+    - Emits a `AlarmNotCanceled` event if the alarm is not canceled successfully
```
[the remark this answers is not in the document]
Comment (Mikefly123) [reply docs-site/components/RtcManager.md:38]: Unrelated to this PR.

File: docs-site/components/RtcManager.md [hunk 6]
```diff
@@ -15,6 +18,32 @@ The RTC Manager component interfaces with the Real Time Clock (RTC) to provide t
...
 
+#### `ALARM_SET` Command Usage
+1. The component is instantiated and initialized during system startup
+2. A ground station sends a `ALARM_SET` command with the desired time
+3. On each command, the component:
+    - Validates that the alarm is at a future date
+    - Emits 'AlarmNotSet' if time is not valid or if an alarm is already present
+    - Sets the time to the RTC alarm if validation passes
+    - Emits a `AlarmSet` event with the previous time if the alarm is set successfully
+    - Emits a `AlarmNotSet` event if the alarm is not set successfully
```
[the remark this answers is not in the document]
Comment (Mikefly123) [reply docs-site/components/RtcManager.md:29]: Unrelated to this PR.

File: docs-site/components/RtcManager.md [hunk 7]
```diff
@@ -15,6 +18,32 @@ The RTC Manager component interfaces with the Real Time Clock (RTC) to provide t
     - Emits a `TimeNotSet` event if the time is not set successfully
     - Emits a `DeviceNotReady` event if the device is not ready
 
+#### `ALARM_SET` Command Usage
```
[the remark this answers is not in the document]
Comment (Mikefly123) [reply docs-site/components/RtcManager.md:21]: Unrelated to this PR.

File: PROVESFlightControllerReference/Components/StartupManager/StartupManager.cpp [hunk 8]
```diff
@@ -186,6 +186,18 @@ void StartupManager ::run_handler(FwIndexType portNum, U32 context) {
         Fw::ParamString first_sequence = this->paramGet_STARTUP_SEQUENCE_FILE(is_valid);
         FW_ASSERT(is_valid == Fw::ParamValid::VALID || is_valid == Fw::ParamValid::DEFAULT);
         this->runSequence_out(0, first_sequence);
+        this->m_transmit_enable_ticks = this->paramGet_TRANSMIT_ENABLE_TICKS(is_valid);
+        FW_ASSERT(is_valid == Fw::ParamValid::VALID || is_valid == Fw::ParamValid::DEFAULT);
+    }
```
[the remark this answers is not in the document]
Comment (Mikefly123) [reply PROVESFlightControllerReference/Components/StartupManager/StartupManager.cpp:263]: I don't think this is actually necessary. The Transmit Enable / Disable state has a compile time default.

File: PROVESFlightControllerReference/Components/StartupManager/docs/sdd.md [hunk 9]
```diff
@@ -106,12 +113,10 @@ Waiting for Quiescence → Running (quiescence period expires OR ARMED=false)
...
-
-### Unit Tests
-
+| REQ-SM-008 | StartupManager shall enable LoRa transmit after `TRANSMIT_ENABLE_TICKS` 1 Hz ticks | Verification: Confirm `HardcodedRadioEnable` and RF transmit after the configured delay |
 
 ## Change Log
 
 | Date | Author | Description |
 |------|--------|-------------|
+| 2026-07-18 | | Document hard-coded radio enable (`TRANSMIT_ENABLE_TICKS`, `enableTransmit`), updated ports/params/events, `/seq/startup.bin` path |
```
[the remark this answers is not in the document]
Comment (Mikefly123) [reply PROVESFlightControllerReference/Components/StartupManager/docs/sdd.md]: @ineskhou recommend that you do indeed reconcile this when you have the chance.

**Review by Mikefly123 — DISMISSED** [review]
I think the core changeset LGTM. 

That being said, I am not a fan of how this PR is packaging many unrelated changes. Our CI is still passing, so I won't block on that fact but we should cleanup the state of our Zephyr Submodule sooner rather than later.

**Comment by ineskhou** [comment]
> I think the core changeset LGTM.
> 
> That being said, I am not a fan of how this PR is packaging many unrelated changes. Our CI is still passing, so I won't block on that fact but we should cleanup the state of our Zephyr Submodule sooner rather than later.

Agreed, I was putting them together in on PR just for the code freeze so I could text things faster. Now that we have more time im downscaling (see the fprime zephyr PR) I will then rerun the soak tests with the 45 min once we're pointing at the right fprime zephyr ref

**Comment by ineskhou** [comment]
DEFAULT_STARTUP_VALUE = 1, TRANSMIT_ENABLE_TICKS = 100
Screenshot 2026-07-22 at 12 43 21 AM


DEFAULT_STARTUP_VALUE = 0, TRANSMIT_ENABLE_TICKS = 100
Screenshot 2026-07-22 at 12 36 14 AM

**Review by hrfarmer — DISMISSED** [review]
Everything here is good to me, but I want to know thoughts on if the `DEFAULT_STARTUP_VALUE == 1` block should also check if the `startup.bin` file exists as well/if the sequence fails to run before running the hardcoded enable, since this is meant to be a fallback for if something happens to the file/filesystem. This is just as a safeguard against potential conflicts if the intended startup sequence has different behavior than the 45min wait before enable, but I don't think that will be the case so probably not an issue.

**Comment by ineskhou** [comment]
@hrfarmer  You provide a good point, which is what if a sequence or an operator command the radio off in order to change the paramaters of the radio

Radio enter safe and the startups sequence both specifically turn the radio off in order to set the satellite back to known paramaters
https://github.com/Open-Source-Space-Foundation/proves-core-reference/blob/main/sequences/radio_enter_safe.seq
https://github.com/Open-Source-Space-Foundation/proves-core-reference/blob/main/sequences/startup.seq

If this port hit while those sequences were running, the radio could be enabled before all the radio parameters were set, potentially putting us in a bad radio state

Possible solutions
- Changing the time on this component to enable the radio much after the startup sequence would, and make sure operators are careful to not run any commands disable. This is the simplest in code by the most complicated in ops
- Have this component check for the startup.seq specifically. This would not include if another sequence turned off the radio though.
- Have this component check for any files existing. This would not account for the operator  commands (see bellow) but would account for any other sequences. But if the startup was not there or not readable and other sequences were 
- Have this component check if a sequence is currently running. This would work under the assumption that we only change the radio using a sequence, which may be incorrect if the file system is buggy. I don't know that we'd want to risk sending a radio disable without a guarantee that we could resent the enable or an ability to get acks to know what commands actually get down from the sat so maybe that is acceptable to write into the ops plan.
- Only run if the radio was never enabled on this boot. This would prevent from operator errors.


My proposal is we combine the last two points. I can add another port in Lora that can tell startup manager if radio was ever enabled on the boot. I will also have startup manager listen to the sequencers run port, to know if any of the three sequences are working. Will put a comment in the toplogy if anyone adds more sequence they need to be considered in startup manager.

**Comment by hrfarmer** [comment]
I think I agree as well with implementing the last two points, but I think additionally, it should work to where if there's a sequence currently running but the radio hasn't been enabled once yet, it will restart the tick timer to check again later instead of cancelling. Miiiight be excessive since its super unlikely it would happen, but technically the safe mode sequence could be running when the timer hits, which would then make the radio not turn on even though the safe mode sequence itself doesn't turn on the radio.

**Comment by ineskhou** [comment]
Agreed about the last one! I was just going to have it wait until the sequence complete came into the port. I'll tag you once I'm done the changes to make sure we capture all the edge cases here

**Comment by ineskhou** [comment]
Testing Procedure
 **- check no safe mode regressions**

✅ Safe Mode Tests all Pass

**This was a test that was ran without running a sequence, to show the normal behavior of the startup component** 
￼
Screenshot 2026-07-24 at 5 41 39 PM

**- Check while sequence is running, component does not turn on** 

Here hardcoded startup doesn’t turn on at 50 seconcs but waits until 110 After the sequence is done to turn on
￼
Screenshot 2026-07-24 at 6 00 30 PM

- Check many sequences are running, radio is only turned on after last sequence completes if the first sequence ends first 


CafCore cmdDins OpCodeCompleted
￼
Ran commands on all 3 sequencers all overlapping, . Hardcoded startup enabled after they were all done
￼
Screenshot 2026-07-24 at 6 14 10 PM
Screenshot 2026-07-24 at 6 13 47 PM

- Check if a sequence turns on the radio, hardcoded startup never goes off

Ran radio_enter_safe, hardcoded startup never turned on after that

Screenshot 2026-07-24 at 6 13 47 PM

￼
- Check if command turned radio on, hardcoded startup does not run
In this run, turned the radio on at 25, and past 50+ the hardcoded startup did not go off
￼
Screenshot 2026-07-24 at 5 56 57 PM
Screenshot 2026-07-24 at 5 57 32 PM

￼
- Check with full 45 min wait
We can see when I set the time to what it is in this PR (2800) 
￼
CotCone omeOio CMO NO 0P completed


- Check 1/0 build flag
All the above were run with a build flag of 1. Ran the code with a build flag of 0 and observed no hardcoded startup:
Screenshot 2026-07-24 at 6 44 42 PM

**Review by collinc04 — COMMENTED** [review]
LGTM

**Review by Mikefly123 — APPROVED** [review]
LGTM! Thanks for all the hard work on this. Hopefully this is an effective flush to gremlins related to file system corruption and the startup sequence.
